### DISCUSSION TITLE
How to safely Exec any (non bubbletea) programs but maintain rendering

### DISCUSSION THREAD

**Discussion by zachatrocity** [charmbracelet/wish#250] [Q&A] [answered]
Hey there! First off, I recently found the charm eco system and I'm blown away and can't believe I didnt find this sooner.

As a learning exercise I built a little Groc API based chat based on the https://github.com/charmbracelet/wish/blob/main/examples/multichat/main.go example and that was working well. I then found  https://github.com/charmbracelet/mods which is exactly what I want in my daily flow, with the exception of accessing through SSH so I dont have to install mods on all my machines and I can centralize my settings in my home server.

This works fairly well and I have a working example here! 

https://git.sr.ht/~zachr/mossh

I wanted this to work in 2 ways:

1. SSH Connection with no passed commands will launch the default mods chat ✅ 
   - `ssh ai.my.domain` or `ssh localhost -p 23234` results in:
       ![image](https://github.com/charmbracelet/wish/assets/6200670/b05d89d2-2913-4828-a938-241e6ddd0e19)

2. SSH Connection with prompt passed as command to the wish server:
   - `ssh ai.my.domain "my sample prompt"` results in:
       ![image](https://github.com/charmbracelet/wish/assets/6200670/bd9f91fc-1b1e-4d10-88bf-7c1daad3650b)

While this works, as you can see from the screenshot all formatting/styling/lipgloss coming from mods isn't applied. I suspect this is because for `#1` I'm using a bubbletea middleware and using tea.Exec like the examples found here https://github.com/charmbracelet/wish/blob/main/examples/wish-exec/main.go

However with option 2, in order to automatically pass the commands as args to mods I had to use a custom middleware as you can see here in my sample code:

```
         func(next ssh.Handler) ssh.Handler {
		return func(sess ssh.Session) {

			// todo: figure out why the renderer wont work through ssh
			// renderer := bubbletea.MakeRenderer(sess)

			if len(sess.Command()) > 0 {
				// commands or args provided pass it to mods
				mods := exec.Command("mods", sess.Command()...)
				mods.Stdout = sess
				mods.Stderr = sess.Stderr()

				mods.Run()

				_ = sess.Exit(1)
			}

			next(sess)
		}
	},
``` 

So my question is this, is it possible to maintain the bubbletea formatting from an application installed on the SSH host (mods) when using `exec.Command`?

I've spent quite some time looking at the examples and source code but I can't quite figure out where I'm going wrong. Any suggestions/tips or am I missing something obvious? 

Here is an example of what mods returns with a similar prompt directly and not through my wish server:

![image](https://github.com/charmbracelet/wish/assets/6200670/e1b0867d-1351-4552-bb40-d764b092d024)

Notice the pretty colors.

**Comment by caarlos0** [comment, marked as answer]
Hey, I believe that if you use `wish.Command` it should work.

wish.Command will create a PTY and run the command within it. 

That said, you can probably remove the that custom handler entirely, you just need to `ssh -t` (to request a tty) and then the original bubbletea handler will do the trick :)

```sh
ssh -t -p 23234 localhost -- hello world
```

Hope that helps.

---

```patch
diff --git a/main.go b/main.go
index ed80a5f..0642607 100644
--- a/main.go
+++ b/main.go
@@ -5,7 +5,6 @@ import (
 	"errors"
 	"net"
 	"os"
-	"os/exec"
 	"os/signal"
 	"syscall"
 	"time"
@@ -48,26 +47,6 @@ func main() {
 				}, []tea.ProgramOption{tea.WithAltScreen()}
 			}),
 			activeterm.Middleware(),
-			func(next ssh.Handler) ssh.Handler {
-				return func(sess ssh.Session) {
-					// todo: figure out why the renderer wont work through ssh
-					// renderer := bubbletea.MakeRenderer(sess)
-
-					if len(sess.Command()) > 0 {
-						// commands or args provided pass it to mods
-						mods := exec.Command("mods", sess.Command()...)
-						mods.Stdout = sess
-						mods.Stderr = sess.Stderr()
-
-						mods.Run()
-
-						_ = sess.Exit(1)
-					}
-
-					next(sess)
-				}
-			},
-
 			logging.Middleware(),
 		),
 	)
```

**Comment by zachatrocity** [comment]
omg i knew it was something silly. Thats amazing, thank you!!

Love this toolset, thanks for all you do.
